Upgrade dependencies and modernize codebase

alexwaibel opened the pull request:
The motivation behind this PR is to resolve dependency security warnings and enable modern JavaScript syntax for the project. This is a pretty large PR, but all of the changes are relatively simple. I've done a few key things here:

1. Move the config files into their own top level `config` directory and renamed since `config` is now encoded in the path. I also moved the remaining source files into a `source` directory so they're not mixed in with things like docker files and other clutter.

1. Upgrade standard to the latest version. This required a node version bump, but added some more rules. I fixed all the new style warnings (let could be const, spaces around object literal brackets, etc).

1. Upgrade Jest to the latest version. The previous one was throwing a lot of security vulnerabilities and it's helpful to be on a recent version of this.

1. Add babel to enable modern syntax. I updated all of the imports to use import/export syntax. I updated the `start` script to automatically compile and serve and added a `start:dev` script with nodemon for development purposes.

1. Upgrade convict to the latest version to resolve further security vulnerabilities.

There's still one security vulnerability with our rate limiter, but I will fix have to fix that in another PR. Like I said before, this PR is pretty large so please ask if anything is unclear.

alexwaibel committed:
Rework project structure so source and config are separate

alexwaibel committed:
Upgrade Standard to latest and resolve errors

alexwaibel committed:
Upgrade jest to latest version

alexwaibel committed:
Initial babel support

alexwaibel committed:
Start converting some modules to ES5

alexwaibel committed:
Update another batch of lib files to ES6

alexwaibel committed:
All lib files and their tests updated to es6

alexwaibel committed:
Convert controllers and thier tests to es6

alexwaibel committed:
Convert remaining import statements to es6

alexwaibel committed:
Utilize spread operator in place of object assign

alexwaibel committed:
Bump convict version to latest to resolve security issue.

alexwaibel committed:
Merge branch 'dev' into aw-upgrade-jest

alexwaibel committed:
Regenerate package lockfile

alexwaibel committed:
Add babel build steps for non-dev envs

alexwaibel committed:
Bump travis node version

eduardoboucas commented on source/lib/ErrorHandler.js:
```diff
@@ -24,12 +22,12 @@ class ErrorHandler {
-      'GITHUB_AUTH_TOKEN_MISSING': 'The site requires a valid GitHub authentication token to be supplied in the `options[github-token]` field',
-      'MISSING_CONFIG_BLOCK': 'Error whilst parsing Staticman config file'
+      RECAPTCHA_MISSING_CREDENTIALS: 'Missing reCAPTCHA API credentials',
+      ECAPTCHA_FAILED_DECRYPT: 'Could not decrypt reCAPTCHA secret',
```
Should this be `ECAPTCHA_FAILED_DECRYPT`?

alexwaibel replied:
Good catch, I resolved this in the latest revision

eduardoboucas reviewed (commented):
Looks great! 😍

Added just a tiny comment to ensure we haven't misnamed a constant.

alexwaibel committed:
Fix error key typo.

alexwaibel committed:
Merge branch 'dev' into aw-upgrade-jest

hispanic commented:
I've got a couple questions:
1. Do you (@alexwaibel or @eduardoboucas) have other refactorings coming down the pipe in the near-to-intermediate time frame (other than https://github.com/eduardoboucas/staticman/pull/397)?
2. Is Alex's recent refactoring and updating being done in preparation for some coming **functionality** changes of significance?

If the answer to either is "yes", would you be up for sharing some details? Looking for some visibility regarding potential impacts to my pull requests. Thanks.

hispanic commented:
I now see Alex's `aw-integration-tests` branch, which I suppose answers my first question (with a "yes"). But, please LMK about the functionality plans. Thanks.

alexwaibel commented:
@hispanic hope this hasn't caused too much confusion, hopefully I can clarify a bit. These recent PRs weren't part of a planned refactoring, I just happened to have some free time this last week. The PR descriptions should give a good overview of the changes but feel free to follow up with any questions. No functionality changed, we just changed the codebase to use some more modern JS features.

> 1. Do you (@alexwaibel or @eduardoboucas) have other refactorings coming down the pipe in the near-to-intermediate time frame (other than #397)?

You figured it out, I'm working on adding some more integration tests in that `aw-integration-tests` branch. I've been refactoring some of the test code as needed. Mainly:

1. Split out acceptance tests into separate files for each controller instead of one large API acceptance test file.
1. Introducing modern JS features to our test code (Reducing nesting by switching from promises to async/await, ditching `require` statements where we can, etc).
1. General test code cleanup (refactoring duplicated code into helper methods, using the more modern jest expectations and mocks, etc).

> 2. Is Alex's recent refactoring and updating being done in preparation for some coming **functionality** changes of significance?

Generally I am **not** looking to introduce any new Staticman functionality. Before adding new functionality to Staticman I think it's more important to address tech debt and make sure all the existing functionality is working as expected and is fully documented. You can take a look at the Projects board for some of the items I hope to tackle. Right now I mainly want to ensure we have a robust test suite so that we can refactor and eventually add features without worry.

hispanic commented:
Thanks for the explanation. I appreciate it. Sounds like a solid roadmap. I'm going to back burner my contribution activity for the foreseeable future. Not a good idea to try and put on an addition while the foundation is being re-poured.

alexwaibel commented:
I'd like to clarify a few things. Hopefully this information will be useful for those wondering about the project state.

> Not a good idea to try and put on an addition while the foundation is being re-poured.

All the big changes are merged into dev now. I'm still working on the new tests but those won't require and Staticman logic changes. My recent PRs, mostly just updated JS syntax, no real logic changes there either. I try to push my branches frequently so others can see what upcoming changes may look like. The [GitHub compare feature](https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/comparing-commits) is quite useful for comparing branches to dev. I'm still happy to accept PRs, even while there's upcoming changes.

> Generally I am not looking to introduce any new Staticman functionality.

To clarify, when I say this, I mean for example I would probably not accept something like adding a third git service provider. I just want to avoid expanding the scope of this project further before I've documented and tested what we currently offer. Your pull requests to bring the webhook endpoint up to speed with the rest of the API design is an example of something that I absolutely would still accept a PR for.

Any deficiencies in the current set of features should certainly still be addressed. Always happy to accept PRs and issues for these kinds of things. When in doubt, you can always open an issue/feature request and I'll let you know if it's in scope.
